csrc/scheduler/hopper_multi_matmul.cpp

```diff
@@ -0,0 +1,1524 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <device_lower/analysis/circular_buffer.h>
+#include <disjoint_set.h>
+#include <id_model/schedule.h>
+#include <instrumentation.h>
+#include <ir/utils.h>
+#include <multidevice/utils.h>
+#include <scheduler/debug_utils.h>
+#include <scheduler/hopper_multi_matmul.h>
+#include <scheduler/matmul.h>
+#include <scheduler/matmul_utils.h>
+#include <scheduler/mma_utils.h>
+#include <scheduler/tools/abstract_tensor.h>
+#include <scheduler/tools/inlining.h>
+#include <scheduler/utils.h>
+#include <val_graph.h>
+#include <val_graph_visitor.h>
+
+// NOTE: included to avoid compilation error caused by missing destructor in
+// 'SchedulerRuntimeInfo'
+#include <runtime/executor_utils.h>
+#include "mma_type.h"
+
+namespace nvfuser {
+
+namespace {
+
+// Returns true if given number is power of 2
+constexpr bool isPowOf2(int64_t x) {
+  return x > 1 && (x & (x - 1)) == 0;
+}
+
+inline IterDomain* representativeId(const AbstractId& abs_id) {
+  if (abs_id.is<IterDomain*>()) {
+    return abs_id.as<IterDomain*>();
+  }
+  NVF_ERROR(abs_id.is<ValGroupAndItsGraph>());
+  return representativeId(abs_id.as<ValGroupAndItsGraph>().group);
+}
+
+// Utility to check concrete static size
+inline void checkConcreteStaticDim(const AbstractId& abs_id) {
+  IterDomain* id = representativeId(abs_id);
+  NVF_ERROR(
+      !id->isBroadcast() && !id->isReduction(),
+      "no support for reduction or broadcast domains, but got ",
+      id->toString());
+  NVF_ERROR(
+      id->extent()->isConstInt(),
+      "swizzled dimension's extend must be known during scheduling, got ",
+      id->toString());
+}
```

Comment:
If this is not needed for TMA swizzle, we can remove it. Otherwise it can go to `mma_utils.cpp`

Reply:
I think the general swizzle logic can go in `mma_utils.cpp` because we need swizzle support for regular smem epilogue.